mobile_verifier/src/reward_share.rs

```diff
@@ -0,0 +1,297 @@
+
+        while let Some(Ok(msg)) = stream.next().await {
+            // NOTE: This will early exit with an error if we fail to decode
+
+            if let Ok(cell_heartbeat_req_v1) = CellHeartbeatReqV1::decode(msg.clone()) {
+                shares.gather_heartbeat(cell_heartbeat_req_v1, after_utc, before_utc)
+            } else if let Ok(speedtest_req_v1) = SpeedtestReqV1::decode(msg) {
+                shares.gather_speedtest(speedtest_req_v1, after_utc, before_utc)
+            } else {
+                continue;
+            }
+        }
+        Ok(shares)
+    }
+
+    fn gather_speedtest(&mut self, speedtest_req: SpeedtestReqV1, after_utc: u64, before_utc: u64) {
+        // TODO: Better error handling for bad public keys
+        let mut speed_share = match SpeedShare::try_from(speedtest_req) {
+            Ok(share) => share,
+            Err(_) => return,
+        };
+
+        if speed_share.timestamp < after_utc || speed_share.timestamp >= before_utc {
+            speed_share.validity = ShareValidity::HeartbeatOutsideRange;
+            self.invalid_speed_shares.push(speed_share);
+        } else {
+            self.speed_shares
+                .entry(speed_share.pub_key.clone())
+                .or_default()
+                .push(speed_share);
+            self.speed_shares_moving_avg.update(&self.speed_shares)
+        }
+    }
+
+    fn gather_heartbeat(
+        &mut self,
+        cell_heartbeat_req_v1: CellHeartbeatReqV1,
+        after_utc: u64,
+        before_utc: u64,
+    ) {
+        let CellHeartbeatReqV1 {
+            pub_key: hb_pub_key,
+            cbsd_id: hb_cbsd_id,
+            timestamp: hb_timestamp,
+            ..
+        } = cell_heartbeat_req_v1;
+
+        if let Some(cell_type) = CellType::from_cbsd_id(&hb_cbsd_id) {
+            // TODO: Will only get inserted in invalid if cbsd_id is a valid cell_type
+            if hb_timestamp < after_utc || hb_timestamp >= before_utc {
+                self.invalid_shares.push(ShareProto {
+                    cbsd_id: hb_cbsd_id,
+                    timestamp: hb_timestamp,
+                    pub_key: hb_pub_key.to_vec(),
+                    weight: crate::bones_to_u64(cell_type.reward_weight()),
+                    cell_type: cell_type as i32,
+                    validity: ShareValidity::HeartbeatOutsideRange as i32,
+                });
+                return;
+            }
+
+            if let Ok(gw_pubkey) = PublicKey::try_from(hb_pub_key.as_ref()) {
+                let share = Share::new(
+                    hb_timestamp,
+                    gw_pubkey,
+                    cell_type.reward_weight(),
+                    cell_type,
+                    ShareValidity::Valid,
+                );
+
+                if self
+                    .shares
+                    .get(&hb_cbsd_id)
+                    .map_or(false, |found_share| found_share.timestamp > hb_timestamp)
+                {
+                    return;
+                }
```

Comment:
do we want to continue to _only_ care about the most recent timestamp here if we're now requiring 2-3 heartbeats within the 24 hour period? do we need a count of valid heartbeats or is this just for reference?

Reply:
I don't think we _will_, but I think the plan was to address that issue after merging this PR. We will need a count of valid heartbeats